pkg/userdata/manager/plugin.go

```diff
@@ -0,0 +1,155 @@
+//
+// UserData plugin manager.
+//
+
+package manager
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	"github.com/kubermatic/machine-controller/pkg/userdata/plugin"
+)
+
+const (
+	// envPluginDir names the environment variable containing
+	// a user defined location of the plugins.
+	envPluginDir = "MACHINE_CONTROLLER_USERDATA_PLUGIN_DIR"
+
+	// pluginPrefix has to be the prefix of all plugin filenames.
+	pluginPrefix = "machine-controller-userdata-"
+)
+
+// Plugin looks for the plugin executable and calls it for
+// each request.
+type Plugin struct {
+	os      providerconfig.OperatingSystem
```

Comment:
Why have the `os` be part of the Plugin? The plugins are stored in a hashmap whose key is the OS so this is redundant